src/wp-includes/global-styles-and-settings.php

```diff
@@ -350,27 +350,34 @@ static function ( $item ) {
  * @return bool Returns true if theme or its parent has a theme.json file, false otherwise.
  */
 function wp_theme_has_theme_json() {
-	static $theme_has_support = null;
+	static $theme_has_support         = null;
+	static $prev_stylesheet_directory = null;
+	static $prev_template_directory   = null;
+
+	$stylesheet_directory = get_stylesheet_directory();
+	$template_directory   = get_template_directory();
+
+	// Make sure that the cached $theme_has_support value is reset when the theme changes.
+	if ( null !== $theme_has_support && (
+		$stylesheet_directory !== $prev_stylesheet_directory ||
+		$template_directory !== $prev_template_directory
+	) ) {
+		$theme_has_support = null;
+	}
+	$prev_stylesheet_directory = $stylesheet_directory;
+	$prev_template_directory   = $template_directory;
 
 	if (
 		null !== $theme_has_support &&
 		/*
 		 * Ignore static cache when the development mode is set to 'theme', to avoid interfering with
 		 * the theme developer's workflow.
 		 */
-		wp_get_development_mode() !== 'theme' &&
-		/*
-		 * Ignore cache when automated test suites are running. Why? To ensure
-		 * the static cache is reset between each test.
-		 */
-		! ( defined( 'WP_RUN_CORE_TESTS' ) && WP_RUN_CORE_TESTS )
```

Comment:
Removing this would be great, but I want to make sure switching the theme in tests is the only reason why this was added in the first place?

Reply:
It seems it was due to the comment right before.

Reply:
I mean, tests all pass with the check removed now. If I comment-out the invalidation condition, then they fail again.

Reply:
Tests were breaking in the AMP plugin due to the lack of cache invalidation when we switch themes in tests: https://github.com/ampproject/amp-wp/pull/7560#discussion_r1256146000

Reply:
The first test to fail is:

```
1) Test_Block_Supports_Layout::test_outer_container_not_restored_for_aligned_image_block_with_themejson_theme
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'<figure class="wp-block-image alignright size-full is-style-round my-custom-classname"><img src="/my-image.jpg"/></figure>'
+'<div class="wp-block-image is-style-round my-custom-classname"><figure class="alignright size-full"><img src="/my-image.jpg"/></figure></div>'

/var/www/tests/phpunit/tests/block-supports/layout.php:168
```

As I understand, it fails because the test tries to switch to a block theme, but `wp_restore_image_outer_container()` includes a check for `wp_theme_has_theme_json()` which returns the cached value for the non-block theme during the unit test bootstrapping.